src/pybamm/expression_tree/operations/serialise.py

```diff
@@ -237,6 +241,228 @@ def load_model(
+            generated from the model name and current datetime.
+
+        Example
+        -------
+        >>> import pybamm
+        >>> model = pybamm.lithium_ion.BasicDFN()
+        >>> from pybamm.expression_tree.operations.serialise import Serialise
+        >>> Serialise.save_custom_model(model, "basicdfn_model")
+
+        """
+        SCHEMA_VERSION = "1.0"
+        model_json = {
+            "pybamm_version": pybamm.__version__,
+            "schema_version": SCHEMA_VERSION,
+            "name": getattr(model, "name", "unnamed_model"),
+            "options": getattr(model, "options", {}),
+            "rhs": [
+                (
+                    Serialise.convert_symbol_to_json(variable),
+                    Serialise.convert_symbol_to_json(rhs_expression),
+                )
+                for variable, rhs_expression in getattr(model, "rhs", {}).items()
+            ],
+            "algebraic": [
+                (
+                    Serialise.convert_symbol_to_json(variable),
+                    Serialise.convert_symbol_to_json(algebraic_expression),
+                )
+                for variable, algebraic_expression in getattr(
+                    model, "algebraic", {}
+                ).items()
+            ],
+            "initial_conditions": [
+                (
+                    Serialise.convert_symbol_to_json(variable),
+                    Serialise.convert_symbol_to_json(initial_value),
+                )
+                for variable, initial_value in getattr(
+                    model, "initial_conditions", {}
+                ).items()
+            ],
+            "boundary_conditions": [
+                (
+                    Serialise.convert_symbol_to_json(variable),
+                    {
+                        side: [
+                            Serialise.convert_symbol_to_json(expression),
+                            boundary_type,
+                        ]
+                        for side, (expression, boundary_type) in conditions.items()
+                    },
+                )
+                for variable, conditions in getattr(
+                    model, "boundary_conditions", {}
+                ).items()
+            ],
+            "events": [
+                {
+                    "name": event.name,
+                    "expression": Serialise.convert_symbol_to_json(event.expression),
+                    "event_type": event.event_type,
+                }
+                for event in getattr(model, "events", [])
+            ],
+            "variables": {
+                str(variable_name): Serialise.convert_symbol_to_json(expression)
+                for variable_name, expression in getattr(model, "variables", {}).items()
+            },
+        }
+
+        if filename is None:
+            filename = model.name + "_" + datetime.now().strftime("%Y_%m_%d-%p%I_%M")
+
+        filename = re.sub(r'[<>:"/\\|?*\x00-\x1F]', "", filename)
```

Comment:
Thanks for improving this. Two things here:
- we should use the 24-hour format instead of the 12-hour format (easier to read and more _computer_-ish)
- the regex removes invalid characters, but it doesn't replace them with safe ones.

I have a suggestion as follows. Please feel free to apply it or iterate on it further:

```suggestion
        if filename is None:
            safe_name = re.sub(r'[^\w\-_.]', '_', model.name or 'unnamed_model')
            timestamp = datetime.now().strftime("%Y_%m_%d_%H_%M_%S")
            filename = f"{safe_name}_{timestamp}"
            
        else:
            # Clean provided filename
            filename = re.sub(r'[<>:"/\\|?*\x00-\x1F]', '_', filename)
```

Then, please also update the test case to match this. Thanks!